### PR TITLE
Fix BranchChildren/Parentage problems that occur with SubProcess and EDAlias

### PR DESCRIPTION
**Pull request by wddgit**
Also adds unit tests to cover these cases.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
Please test

File: DataFormats/Provenance/interface/Provenance.h [hunk 1]
```diff
@@ -45,6 +45,7 @@ namespace edm {
 
     ProductProvenance const* productProvenance() const;
     BranchID const& branchID() const {return stable().branchID();}
+    BranchID const& originalBranchID() const {return stable().originalBranchID();}
```
Comment (Dr15Jones) [inline DataFormats/Provenance/interface/Provenance.h]: Please add a comment explaining the difference between `branchID` and `originalBranchID`

File: FWCore/Framework/src/ScheduleItems.cc [hunk 2]
```diff
@@ -36,6 +38,7 @@ namespace edm {
       preg_(new SignallingProductRegistry(preg)),
       branchIDListHelper_(new BranchIDListHelper),
       thinnedAssociationsHelper_(new ThinnedAssociationsHelper),
+      subProcessParentageHelper_(new SubProcessParentageHelper),
```
Comment (Dr15Jones) [inline FWCore/Framework/src/ScheduleItems.cc]: How about switching to `std::make_shared<>()`? That is actually more memory efficient

File: FWCore/Framework/src/ScheduleItems.cc [hunk 3]
```diff
@@ -141,6 +144,7 @@ namespace edm {
                      *preg_,
                      *branchIDListHelper_,
                      *thinnedAssociationsHelper_,
+                     subProcessParentageHelper_ ? &*subProcessParentageHelper_ : nullptr,
```
Comment (Dr15Jones) [inline FWCore/Framework/src/ScheduleItems.cc]: probably simpler to do `subProcessParentageHelper_ ? subProcessParentageHelper_.get() : nullptr`

File: IOPool/Input/src/RootFile.h [hunk 4]
```diff
@@ -227,7 +227,7 @@ namespace edm {
     std::shared_ptr<LuminosityBlockAuxiliary> fillLumiAuxiliary();
     std::shared_ptr<RunAuxiliary> fillRunAuxiliary();
     std::string const& newBranchToOldBranch(std::string const& newBranch) const;
-    void markBranchToBeDropped(bool dropDescendants, BranchID const& branchID, std::set<BranchID>& branchesToDrop) const;
+    void markBranchToBeDropped(bool dropDescendants, BranchDescription const& branchy, std::set<BranchID>& branchesToDrop, std::map<BranchID, BranchID> const& droppedToKeptAlias) const;
```
Comment (Dr15Jones) [inline IOPool/Input/src/RootFile.h]: `branchy` seems like an odd name

**Review by Dr15Jones — CHANGES_REQUESTED** [review]
Done with my review.

**Comment by wddgit** [comment]
@Dr15Jones I agree with all your comments and submitted the appropriate modifications in the last commit I pushed.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by davidlange6** [comment]
-1
test failure looks related.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
@davidlange6 as far as David Dagenhart and I can tell, the unit test failure is from the problem fixed by #17946 and not from this pull request.

**Comment by davidlange6** [comment]
please test

ok, lets test again

**Comment by davidlange6** [comment]
+1

**Comment by dmitrijus** [comment]
Hi @wddgit, can this be backported to 90x?

This problem causes a segfault in IOPool/Streamer modules, this PR fixes it.

**Comment by wddgit** [comment]
@dmitrijus We can probably backport the change, but we are interested in understanding how and why the seg fault is occurring. Can you tell me how to reproduce it so I can investigate this? If you already know exactly how the seg fault is occurring, can you explain it? Thanks.

**Comment by dmitrijus** [comment]
@wddgit 
To reproduce merge https://github.com/cms-sw/cmssw/pull/18066 and use 
cmsRun DQMStreamerOutputRepackerTest_cfg.py <any_root_with_mc_data>.root 

The crash is a (double) null-pointer dereference in this line,
https://github.com/cms-sw/cmssw/blob/master/IOPool/Streamer/src/StreamSerializer.cc#L149

**Comment by wddgit** [comment]
@Dr15Jones I reproduced the problem. If you write a file containing an EDAlias to a ROOT format output file and then in a subsequent process you read that file and try to run the streamer output module, then a seg fault occurs. The problem is indeed in the line he pointed to in StreamSerializer.cc. It successfully get the product using the EDAlias, then using the EDAlias BranchID tries to get the ProductProvenance, but that is stored keyed on the original BranchID not the EDAlias BranchID so it fails and returns a nullptr. The streamer output module does not check the pointer and dereferences and there is a seg fault. This was fixed by this PR in 9_1_X, but not backported to 9_0_X

I reproduced the error running a Framework unit test to make the file with the EDAlias
(cmsRun FWCore/Integration/test/testGetBy1_cfg.py)

and the following to read it and seg fault. It has nothing to do with the DQM PR.

```
import FWCore.ParameterSet.Config as cms

process = cms.Process("PROD2")

process.source = cms.Source("PoolSource",
    fileNames = cms.untracked.vstring(
        'file:testGetBy1.root'
    )
)

process.out = cms.OutputModule("EventStreamFileWriter",
    fileName = cms.untracked.string('testSeriesOfProcessesHLT.dat'),
    compression_level = cms.untracked.int32(1),
    use_compression = cms.untracked.bool(True),
    max_event_size = cms.untracked.int32(7000000)
)

process.e = cms.EndPath(process.out)
```
Should I backport the entire PR (and the other related PR that followed) to 9_0_X? This has probably been there for a while. It is somewhat unusual to run streamer output on a file containing an EDAlias and that is probably why no one noticed before.

**Comment by wddgit** [comment]
I did not try it but this might also fail if you dropped the parentage and tried to run streamer output. So it could be looked at as also a bug in streamer output.

**Comment by Dr15Jones** [comment]
Let's start with a fix to the Streamer. Both in 9_1 and 9_0.

**Comment by wddgit** [comment]
I'm working on fixing the Streamer IO. It actually is not built to handle a nullptr for the ProductProvenance at all so this is more than a one or two line fix. I'll need to trace this all the way through streamer input and output and then make several little fixes in the entire streamer input and output chain to make it work.

**Comment by dmitrijus** [comment]
@wddgit Thanks!

**Comment by wddgit** [comment]
@dmitrijus I submitted pull requests for the Streamer Output Module problem. PR #18126 was already merged into 9_1_X and PR #18135 is the back port to 9_0_X which was just submitted (not merged yet). These should be sufficient to avoid the seg fault you saw.
@Dr15Jones My understanding is that we are not backporting this PR because we want to maintain stability in 9_0_X, avoid the effort to resolve conflicts, and the risk associated with it. Let me know if you want me to backport this also.

**Pull request merged**